controllers/tenant_controller.go

```diff
@@ -0,0 +1,528 @@
+			continue
+		}
+		err := r.disownNamespace(ctx, &ns)
+		if err != nil {
+			return err
+		}
+		err = r.removeRBAC(ctx, tenant, &ns)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *TenantReconciler) reconcileArgoCD(ctx context.Context, tenant *cattagev1beta1.Tenant) error {
+	logger := log.FromContext(ctx)
+
+	orig := argocd.AppProject()
+	err := r.client.Get(ctx, client.ObjectKey{Namespace: r.config.ArgoCD.Namespace, Name: tenant.Name}, orig)
+	if err != nil && !apierrors.IsNotFound(err) {
+		logger.Error(err, "failed to get AppProject")
+		return err
+	}
+
+	nss := &corev1.NamespaceList{}
+	if err := r.client.List(ctx, nss, client.MatchingFields{constants.TenantNamespaces: tenant.Name}); err != nil {
+		return fmt.Errorf("failed to list namespaces: %w", err)
+	}
+	namespaces := make([]string, len(nss.Items))
+	for i, ns := range nss.Items {
+		namespaces[i] = ns.Name
+	}
+
+	tpl, err := template.New("AppProject Template").Parse(r.config.ArgoCD.AppProjectTemplate)
+	if err != nil {
+		return err
+	}
+
+	var buf bytes.Buffer
+	err = tpl.Execute(&buf, struct {
+		Name         string
+		Namespaces   []string
+		Roles        map[string][]string
+		Repositories []string
+	}{
+		Name:         tenant.Name,
+		Namespaces:   namespaces,
+		Roles:        rolesMap(tenant.Spec.Delegates),
+		Repositories: tenant.Spec.ArgoCD.Repositories,
+	})
+	if err != nil {
+		return err
+	}
+
+	proj := argocd.AppProject()
+	dec := yaml.NewDecodingSerializer(unstructured.UnstructuredJSONScheme)
+	_, _, err = dec.Decode(buf.Bytes(), nil, proj)
+	if err != nil {
+		return err
+	}
+
+	proj.SetNamespace(r.config.ArgoCD.Namespace)
+	proj.SetName(tenant.Name)
+	proj.SetLabels(map[string]string{
+		constants.OwnerTenant: tenant.Name,
+	})
+
+	managed, err := extract.ExtractManagedFields(orig, constants.FieldManager)
+	if err != nil {
+		return err
+	}
+	if equality.Semantic.DeepEqual(proj, managed) {
+		return nil
+	}
+
+	err = r.client.Patch(ctx, proj, client.Apply, &client.PatchOptions{
+		Force:        pointer.BoolPtr(true),
+		FieldManager: constants.FieldManager,
+	})
+	if err != nil {
+		logger.Error(err, "failed to patch AppProject")
+		return err
+	}
+	logger.Info("AppProject successfully reconciled")
+
+	return nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *TenantReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	tenantHandler := func(o client.Object, q workqueue.RateLimitingInterface) {
+		owner := o.GetLabels()[constants.OwnerTenant]
+		if owner == "" {
+			return
+		}
+		q.Add(reconcile.Request{NamespacedName: types.NamespacedName{
+			Name: owner,
+		}})
+	}
+
+	funcs := handler.Funcs{
+		CreateFunc: func(ev event.CreateEvent, q workqueue.RateLimitingInterface) {
+			tenantHandler(ev.Object, q)
+		},
+		UpdateFunc: func(ev event.UpdateEvent, q workqueue.RateLimitingInterface) {
+			if ev.ObjectNew.GetDeletionTimestamp() != nil {
+				return
+			}
+			tenantHandler(ev.ObjectOld, q)
```

Comment:
We may add `cattage.cybozu.io/tenant` label to namespaces adversarially and it does not trigger reconciliation. Is it OK?